StyleCop.Analyzers/StyleCop.Analyzers/GeneratedCodeAnalysisExtensions.cs

```diff
@@ -0,0 +1,97 @@
+﻿// This file originally obtained from 
+// https://raw.githubusercontent.com/code-cracker/code-cracker/08c1a01337964924eeed12be8b14c8ce8ec6b626/src/Common/CodeCracker.Common/Extensions/GeneratedCodeAnalysisExtensions.cs
+// It is subject to the Apache License 2.0
+// This file has been modified since obtaining it from its original source.
+
+namespace StyleCop.Analyzers
+{
+    using System.Linq;
+    using System.Runtime.CompilerServices;
+    using System.Text.RegularExpressions;
+    using System.Threading;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    internal static class GeneratedCodeAnalysisExtensions
+    {
+        /// <summary>
+        /// A cache of the result of computing whether a document has an auto-generated header.
+        /// </summary>
+        /// <remarks>
+        /// This allows many analyzers that run on every token in the file to avoid checking
+        /// the same state in the document repeatedly.
+        /// </remarks>
+        private static readonly ConditionalWeakTable<SyntaxTree, StrongBox<bool?>> GeneratedHeaderPresentCheck = new ConditionalWeakTable<SyntaxTree, StrongBox<bool?>>();
+
+        internal static bool IsGenerated(this SyntaxNodeAnalysisContext context, CancellationToken cancellationToken) => (context.Node.SyntaxTree?.IsGenerated(cancellationToken) ?? false) || (context.Node?.IsGenerated() ?? false);
```

Comment:
:exclamation: I don't like how there are several `IsGenerated` methods, but they all have completely different behavior. In particular, the `SyntaxNode` method only looks at attributes, but the `SyntaxTree` method only looks at the file path and leading comment(s).